compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/JvmBackendContext.kt

```diff
@@ -71,6 +73,13 @@ class JvmBackendContext(
         localClassInfo[container.attributeOwnerId] = value
     }
 
+    // TODO cache these at ClassCodegen level. Currently, sharing this map between classes in a module is required
+    //      because IrSourceCompilerForInline constructs a new (Fake)ClassCodegen for every call to
+    //      an inline function in the same module. Thus, if two inline functions happen to have the same name
+    //      and call a third inline function that has an anonymous object, the one which is called last
+    //      will overwrite the other's regenerated copy. (Or don't recompile the inline function for every call.)
+    internal val regeneratedObjectNameGenerators = mutableMapOf<Pair<IrClass, Name?>, NameGenerator>()
```

Comment:
Would such logic stable in case of incremental compilation?

Reply:
I believe so. If the file containing the overloaded inline functions (`sameName` in the test case) is recompiled, then so should be all the use sites; all regenerated classes will be created again with the same names as in a full compilation scenario. If the file is not recompiled, then call sites will read bytecode for the inline functions from the class file instead of going through FakeClassCodegen, and no potentionally clashing regenerations happen in the first place.